python/lsst/daf/butler/registry/interfaces/database.py

```diff
@@ -0,0 +1,927 @@
+            The names of columns that will be used to constrain the rows to
+            be deleted; these will be combined via ``AND`` to form the
+            ``WHERE`` clause of the delete query.
+        rows
+            Positional arguments are the keys of rows to be deleted, asIterable
+        Return
+        ------
+        count : `int`
+            Number of rows deleted.
+
+        Raises
+        ------
+        ReadOnlyDatabaseError
+            Raised if `isWriteable` returns `False` when this method is called.
+
+        Notes
+        -----
+        May be used inside transaction contexts, so implementations may not
+        perform operations that interrupt transactions.
+
+        The default implementation should be sufficient for most derived
+        classes.
+        """
+        if not self.isWriteable():
+            raise ReadOnlyDatabaseError(f"Attempt to delete from read-only database '{self}'.")
+        sql = table.delete()
+        whereTerms = [table.columns[name] == sqlalchemy.sql.bindparam(name) for name in columns]
+        if whereTerms:
+            sql = sql.where(sqlalchemy.sql.and_(*whereTerms))
+        return self._connection.execute(sql, *rows).rowcount
+
+    def update(self, table: sqlalchemy.schema.Table, keys: Dict[str, str], *rows: dict) -> int:
+        """Update one or more rows in a table.
+
+        Parameters
+        ----------
+        table : `sqlalchemy.schema.Table`
+            Table containing the rows to be updated.
+        keys : `dict` [`str`, `str`]
+            A mapping from the names of columns that will be used to search for
+            existing rows to the keys that will hold these values in ``rows``
+            dictionaries.  Note that these may not be the same due to
+            SQLAlchemy limitations.
+        rows
+            Positional arguments are the rows to be updated, as dictionaries
+            mapping column name to value.  The keys in all dictionaries must
+            be the combination of the names in ``keys`` and ``values``.
+
+        Raises
+        ------
+        ReadOnlyDatabaseError
+            Raised if `isWriteable` returns `False` when this method is called.
+
+        Returns
+        -------
+        count : `int`
+            Number of rows matched (regardless of whether the update actually
+            modified them).
+
+        Notes
+        -----
+        May be used inside transaction contexts, so implementations may not
+        perform operations that interrupt transactions.
+
+        The default implementation should be sufficient for most derived
+        classes.
+        """
+        if not self.isWriteable():
+            raise ReadOnlyDatabaseError(f"Attempt to update read-only database '{self}'.")
+        sql = table.update().where(
+            sqlalchemy.sql.and_(*[table.columns[k] == sqlalchemy.sql.bindparam(v) for k, v in keys.items()])
+        )
+        return self._connection.execute(sql, *rows).rowcount
```

Comment:
I think something is missing here, UPDATE needs `.values()` method too in addition to `.where()`.